### ISSUE TITLE
zfs receive stalls/slowdown with dnodesize!=legacy

### ISSUE THREAD

**Issue by hhhappe** [zfsonlinux/zfs#8458]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| CentOS
Distribution Version	| 7.6
Linux Kernel	| 3.10.0-957.5.1.el7.x86_64
Architecture	| x86_64
ZFS Version	| 0.7.12-1
SPL Version	| 0.7.12-1

### Describe the problem you're observing
zfs receive of fs with dnodesize!=legacy stalls or slows down. Does a lot of IOPS, but gets next to nowhere.

This was discovered receiving a Lustre metadata fs, so I started investigating a bit to reproduce. I was thinking lots of files and found that dnodesize made a difference.

### Describe how to reproduce the problem
Create an fs with dnodesize=auto (also confirmed stall with 4k). Create (touch) a million files in the same directory. send/receive it and it stalls. 'zfs list' shows that the fs uses is 130M. The receive stalls/slowdown at  94M. 

So far so good. Next step is to test older versions of ZFS. Perhaps, how the number of files and dir structure affects the issue.

**Comment by hhhappe** [comment]
Update

It only happens when the sending side has dnodesize!=legacy. It makes no difference if it is received into a dnodesize=legacy fs.

Doing a pool in memory (file in /dev/shm), the write rate just goes up. iostat shows:

```
pool        alloc   free   read  write   read  write
----------  -----  -----  -----  -----  -----  -----
mem          152M  15.7G      0  53.8K      0   209M
```

Had a zfs 0.7.9 system on kernel 2.6.32-754.6.3.el6.x86_64. It did the receive without any problems. The same goes for zfs 0.7.6 on kernel 3.10.0-693.17.1.el7.x86_64.

**Comment by hhhappe** [comment]
Update 2

On the kernel from the initial post 0.7.9 works, while 0.7.11 does not.

**Comment by behlendorf** [comment]
Thank you for narrowing this down.   I suspect that you're being throttled by this 45f0437912f4688bb15b353c8773b0e37e0b1a89 fix which was added in 0.7.11, but is not present in 0.7.9.  The logic which was added handles the case where a given dnode object number needs to be recycled within the send stream, because it was freed and reallocated, but its old and new size are different (i.e 512b -> 1024k).  This should be pretty rare, but when it's encountered it needs to force a transaction group sync which is going to be slow.  This is likely why you're seeing good performance up to 94/130M, then performance drops severely but you still see a large amount of IO.

The key question I have is why exactly we see these mixed sizes for recycled objects in the send stream.  Are you able to reproduce this with a newly created Lustre MDT which has always used a dnodesize=auto?  This may have something to do with how the Lustre MDT is creating the objects.

**Comment by hhhappe** [comment]
Perhaps I wasn't clear. The reproduction tests were done on newly created ZFS posix filesystems.

**Comment by behlendorf** [comment]
Oh I see, would it be possible to post your reproducer?

**Comment by hhhappe** [comment]
Here you go:

```
#!/bin/bash

# Create pool in memory.
truncate -s 1g /dev/shm/zfs.img
zpool create mem /dev/shm/zfs.img

zfs set dnodesize=auto mem

# Create a lot of files
n=0
while [ $n -lt 100000 ]; do
  echo /mem/f$n
  let n++
done|xargs touch

# send to file
zfs snapshot mem@0
zfs send mem@0 > /tmp/mem0

echo Receiving...
cat /tmp/mem0|zfs receive mem/receive
```

Note: 100k files is enough. Think it is more a matter of being able to notice it before it completes. With 1M files it can complete but it takes a lot of time.

**Comment by hhhappe** [comment]
A few more findings.
The time it takes is proportional to the number of files, which is also the case for legacy.
It takes about 25 times longer with auto compared to legacy with in-memory pool. Perhaps, this will hit harder on media that are more IO bound.

**Comment by hhhappe** [comment]
Just gave 0.8.1 a spin. It solves the issue. Tanks.

**Comment by gmelikov** [comment]
Closed in 0.8.1.
